test/hotspot/jtreg/compiler/codegen/BMI2.java

```diff
@@ -0,0 +1,449 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @summary Support BMI2 instructions on x86/x64
+ *
+ * @run main/othervm -Xbatch -XX:-TieredCompilation -XX:+UnlockDiagnosticVMOptions -XX:+AbortVMOnCompilationFailure
+ *      -XX:CompileCommand=compileonly,compiler.codegen.BMI2$BMITests::*
```

Comment:
It might be better to use "-XX:CompileCommand=dontinline,compiler.codegen.BMI2$BMITests::*" Then the main methods can still be compiled. As long as BMI2$BMITests method isn't inlining other BMI2$BMITests methods this is will just work just fine.

Reply:
Sounds reasonable.

Reply:
I tried to use "-XX:CompileCommand=dontinline,compiler.codegen.BMI2$BMITests::*" however I'm unable to directly verify the C2 compilation of bzhiI2L() method and therefor bzhiq instructions that should be generated. I see the testZeroBzhiI2L() testOneBzhiI2L() methods get compiled and they both make a static call to  bzhiI2L() but it doesn't appear to be C2 compiled.

I'm using the following command: 'make test TEST="jtreg:test/hotspot/jtreg/compiler/codegen/BMI2.java" JTREG="VM_OPTIONS=-XX:+UnlockDiagnosticVMOptions -XX:-TieredCompilation -XX:+PrintAssembly -Xbatch -XX:CompileCommand=print,compiler.codegen.BMI2$BMITests::* -XX:CompileCommand=dontinline,compiler.codegen.BMI2$BMITests::*" CONF=linux-x86_64-server-slowdebug'

But I've tried many variations. If I use "-XX:CompileCommand=compileonly,compiler.codegen.BMI2$BMITests::*" I can manually verify that bzhiq is generated by C2 compiler.

I'd prefer to leave this as is unless I'm missing something.

Reply:
Nils's suggestion is reasonable. With "compileonly,compiler.codegen.BMI2$BMITests::*" the test will be slow because BMI2::test*BzhiI2L() methods are not compiled.
 Use -XX:+PrintCompilation to see what is going on with compilations. I think you will see several recompilations until all switch cases are executed. That is why I want loops in test*BzhiI2L() be inverted (ITERATIONS as external loop) - you will get final code version much sooner. You can preserve golden values (first run) in array to compare later.
BTW, you don't need 2 methods - pass `ix` as parameter.

```
public static void testBzhiI2L(int ix) {
       long[] goldv = new long[16];
       for (int i = 0; i <= 15; i++) {
             goldv[i] = BMITests.bzhiI2L(ix, i);
       }
       for (int i2 = 0; i2 < ITERATIONS; i2++) {
            for (int i = 0; i <= 15; i++) {
                 long v = BMITests.bzhiI2L(ix, i);
                 if (v != goldv[i]) {
                         throw new Error(returnBzhiI2LErrMessage (gold[i], v));
                 }
            }
      }
}
```

Reply:
Thanks @vnkozlov and @neliasso. I've made the changes you suggest.

I'm still unsure about `-XX:CompileCommand=dontinline,compiler.codegen.BMI2$BMITests::*` over `-XX:CompileCommand=compileonly,compiler.codegen.BMI2$BMITests::*` but I will trust in your knowledge and experience in this area. It doesn't materially change the test just how it's run by default and the test should C2 compile and compare 
`bzhiq`
version to 
```
andI
movslq 
```

Even  with -XX:+PrintCompilation I'm not able to see this happening in test .jtr using  `-XX:CompileCommand=dontinline,compiler.codegen.BMI2$BMITests::*` and printing assembly, inlining, compilation but perhaps I am reading the logs incorrectly.